L1 renaming at each declaration

tautschnig opened the pull request:
~I still need to review why the second commit is necessary, thus marking work-in-progress.~

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
L1 renaming at each declaration

This was done up until c8f0f7b7e77c6d0bf872ef4e3c91bfa6c5c692b4. The
(observable) problem tackled here is that objects declared in loops got the same
L1 name, which implies they had the same address. Thus objects were spuriously
reported as dead when their address was taken, because a pointer to the latest
(live) object would be the same as one to an earlier (genuinely dead) object.

A side effect of this change is objects generated for test suites now have L1
indices, as witnessed in the pointer-function-parameters test.

tautschnig commented:
Actually marking this RFC: the second commit makes perfect sense, it's odd that previously the suffix was not included. Could TG people (@thk123, @allredj, @peterschrammel, or anyone else) let me know whether this breaks anything on your end?

smowton commented:
Would you mind elaborating briefly what kind of thing previously went wrong? I get the vague idea but just want to check my understanding. Some of that explanation might usefully make its way into comments in `symex_decl`.

Also am I right in thinking the `symex_dead` changes are purely cleanup and don't change behaviour?

tautschnig commented:
Yes, I need to add a lot more documentation here/in the commit message. The (observable) problem is that objects declared in loops got the same L1 name, which implies they had the same address. Thus objects were spuriously reported as dead when their address was taken, because a pointer to the latest (live) object would be the same as one to an earlier (genuinely dead) object. A problem not yet fully fixed is regression/cbmc/vla3, which still fails despite the work here.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 205baa873c770f6f472f882c5d2b7ca834a38fe2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98894303
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig committed:
Clarify comments in symex_dead

Try to make clearer what we can(not) remove from renaming maps. Also do small
cleanups of the code.

tautschnig commented:
I have cleaned up the changes a bit and made the commit messages clearer.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ed73813fa24d0b09e8efd0d31b5ce99de813c6b3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99232393

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 953e2d0705b603cb44e96bbf104677f072e43e9c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99247250

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1bab7c1ca566c4b923f9576b5bfc30d2aedba3d3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99321819

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 83e58f30c673d0d5e49c78ee2f1c36792306f862).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99759647

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 9f90b20692b1d4c941a9a09a22a46924a89d9de1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101116350
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 450aca6d7d4a7dd647c15a95e9d57067f8847074).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101119088

smowton commented:
Had been ignoring this because the description suggests this is still WIP; I now notice that is not the case

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -797,3 +797,34 @@ void goto_statet::output_propagation_map(std::ostream &out)
+  const irep_idt l0_name = ssa.get_identifier();
+
+  // save old L1 name, if any
+  auto c_it = level1.current_names.find(l0_name);
```
Do a single `emplace` and then adjust if it already exists, rather than the find / emplace double-lookup pattern?

romainbrenguier replied:
:+1:

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -797,3 +797,34 @@ void goto_statet::output_propagation_map(std::ostream &out)
+  rename(ssa, ns, goto_symex_statet::L1);
+  const irep_idt &l1_name = ssa.get_identifier();
+
+  // unique -- store
```
This comment is not clear

tautschnig replied:
Resolved as part of the move from a `PRECONDITION` to an `INVARIANT`.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -797,3 +797,34 @@ void goto_statet::output_propagation_map(std::ostream &out)
+
+  // unique -- store
+  if(!frame.local_objects.insert(l1_name).second)
+    PRECONDITION(false);
```
This is not a precondition. Use `CHECK_RETURN` or `INVARIANT` instead

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_decl.cpp:
```diff
@@ -34,11 +34,14 @@ void goto_symext::symex_decl(statet &state)
-  state.rename(ssa, ns, goto_symex_statet::L1);
+  state.rename(ssa, ns, goto_symex_statet::L0);
+  const std::size_t fresh_l1_index =
+    path_storage.get_unique_index(ssa.get_identifier(), 1);
```
Is there a particular meaning for the 1 here?

tautschnig replied:
Comment added. (It's just always been that way, but it's really arbitrary. Changing it to 0 would require updating several tests.)

romainbrenguier commented on src/goto-symex/symex_function_call.cpp:
```diff
@@ -380,69 +381,28 @@ void goto_symext::symex_end_of_function(statet &state)
 
-/// preserves locality of local variables of a given function by applying L1
-/// renaming to the local identifiers
+/// preserves locality of parameters of a given function by applying L1
```
Capitalize

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/symex_function_call.cpp:
```diff
@@ -380,69 +381,28 @@ void goto_symext::symex_end_of_function(statet &state)
   {
     // get L0 name
-    ssa_exprt ssa(ns.lookup(*it).symbol_expr());
+    ssa_exprt ssa(ns.lookup(param.get_identifier()).symbol_expr());
```
brace initialization is preferred

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/path_storage.h:
```diff
@@ -90,11 +90,24 @@ class path_storaget
+    auto entry = unique_index_map.emplace(id, minimum_index);
+
+    if(!entry.second)
+      entry.first->second = std::max(entry.first->second + 1, minimum_index);
```
It looks like we could avoid the `if` condition if the function is always called with the same minimum index, I don't know if that's always the case.

tautschnig replied:
No, even right now we (intentionally) use `0` for L1 indices when setting up new threads, but elsewhere use `1`. And also that facility is rather general, we could as well use it for non-det counters any maybe others.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9806ccdcdd760d9cefa1759bdfa64e5eb7a0ee0b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102906311

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9f9c8b86c7fd9ec557d8578d4ebeeb382786b7a7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102972059
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2777f858e98b16804b1e31604c44d52b80facc73).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102999251
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a38221826bcc839bf229fbd5e2091f22bb538f7a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103000236
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 3dbd7baf4aa9eb116cc36ec73415dae87d89975c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103022135
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d165461a6af5f138c468a9899930ac4a9e5dfd26).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103065221
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

romainbrenguier commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -146,6 +143,10 @@ class goto_symex_statet final : public goto_statet
   }
 
+  /// Turn L0-renamed \p ssa into an L1-renamed SSA expression with an L1 index
+  /// \p l1_index that must not have previously been used.
```
nitpick I think it would be slightly clearer if you cut the sentence in two.

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/path_storage.h:
```diff
@@ -98,11 +98,24 @@ class path_storaget
   virtual patht &private_peek() = 0;
   virtual void private_pop() = 0;
+
+  std::unordered_map<irep_idt, std::size_t> unique_index_map;
```
maybe add some documentation

tautschnig replied:
Done.

romainbrenguier commented on src/goto-symex/path_storage.h:
```diff
@@ -98,11 +98,24 @@ class path_storaget
   virtual patht &private_peek() = 0;
   virtual void private_pop() = 0;
+
+  std::unordered_map<irep_idt, std::size_t> unique_index_map;
```
maybe add some documentation

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a397a58be580dac27a7c9f47b15cab7232a8b0d3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103457649
